source/val/validate_function.cpp

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/val/validate.h"
+
+#include "source/opcode.h"
+#include "source/val/instruction.h"
+#include "source/val/validation_state.h"
+
+namespace spvtools {
+namespace val {
+namespace {
+
+spv_result_t ValidateFunction(ValidationState_t& _, const Instruction* inst) {
+  const auto function_type_id = inst->GetOperandAs<uint32_t>(3);
+  const auto function_type = _.FindDef(function_type_id);
+  if (!function_type || SpvOpTypeFunction != function_type->opcode()) {
+    return _.diag(SPV_ERROR_INVALID_ID, inst)
+           << "OpFunction Function Type <id> '" << _.getIdName(function_type_id)
+           << "' is not a function type.";
+  }
+
+  const auto return_id = function_type->GetOperandAs<uint32_t>(1);
+  if (return_id != inst->type_id()) {
+    return _.diag(SPV_ERROR_INVALID_ID, inst)
+           << "OpFunction Result Type <id> '" << _.getIdName(inst->type_id())
+           << "' does not match the Function Type's return type <id> '"
+           << _.getIdName(return_id) << "'.";
+  }
+  return SPV_SUCCESS;
+}
+
+spv_result_t ValidateFunctionParameter(ValidationState_t& _,
+                                       const Instruction* inst) {
+  // NOTE: Find OpFunction & ensure OpFunctionParameter is not out of place.
+  size_t param_index = 0;
+  size_t inst_num = inst->LineNum() - 1;
+  if (inst_num == 0) {
+    return _.diag(SPV_ERROR_INVALID_LAYOUT, inst)
+           << "Function parameter cannot be the first instruction.";
+  }
+
+  auto func_inst = &_.ordered_instructions()[inst_num];
+  if (func_inst != inst) {
```

Comment:
Is this check needed? Seems like we're just checking if we used the vector correctly?